src/main/java/org/embulk/input/marketo/delegate/ProgramInputPlugin.java

```diff
@@ -0,0 +1,222 @@
+
+        @Config("latest_updated_at")
+        @ConfigDefault("null")
+        Optional<Date> getLatestUpdatedAt();
+
+        @Config("filter_type")
+        @ConfigDefault("null")
+        Optional<String> getFilterType();
+
+        @Config("filter_values")
+        @ConfigDefault("null")
+        Optional<List<String>> getFilterValues();
+
+        @Config("report_duration")
+        @ConfigDefault("null")
+        Optional<Long> getReportDuration();
+
+        void setLatestUpdatedAt(Optional<Date> latestUpdatedAt);
+    }
+
+    public ProgramInputPlugin()
+    {
+    }
+
+    @Override
+    public void validateInputTask(PluginTask task)
+    {
+        super.validateInputTask(task);
+        // validate if query_by is selected
+        if (task.getQueryBy().isPresent()) {
+            switch(task.getQueryBy().get()) {
+            case TAG_TYPE:
+                //make sure tag type and tag value are not empty
+                if (!task.getTagType().isPresent() || !task.getTagVallue().isPresent()) {
+                    throw new ConfigException("tag_type and tag_value are required when query by Tag Type");
+                }
+                break;
+            case DATE_RANGE:
+                // make sure earliest_updated_at is not empty
+                if (!task.getEarliestUpdatedAt().isPresent()) {
+                    throw new ConfigException("`earliest_updated_at` is required when query by Date Range");
+                }
+
+                DateTime earliest = new DateTime(task.getEarliestUpdatedAt().get());
+                // make sure latest_updated_at is not empty
+                if (!task.getLatestUpdatedAt().isPresent()) {
+                    throw new ConfigException("`latest_updated_at` is required when query by Date Range");
+                }
+
+                // If incremental report
+                if (task.getIncremental()) {
+                    // The very first run
+                    if (!task.getReportDuration().isPresent()) {
+                        DateTime latest = new DateTime(task.getLatestUpdatedAt().get());
+                        if (latest.isAfter(DateTime.now())) {
+                            throw new ConfigException(String.format("`latest_updated_at` (%s) cannot precede the current date (%s) when incremental import",
+                                            latest.toString(DATE_FORMATTER),
+                                            (DateTime.now().toString(DATE_FORMATTER))));
+                        }
+                    }
+                    else {
+                        // Update the latestUpdatedAt for the config
+                        DateTime latest = earliest.plus(task.getReportDuration().get());
+                        // Only import until now
+                        if (latest.isAfter(DateTime.now())) {
+                            latest = DateTime.now();
+                        }
+                        task.setLatestUpdatedAt(Optional.of(latest.toDate()));
```

Comment:
I'm little confused here, `latestUpdatedAt` could be updated by using `earliest` and `report duration`, it means there is still acceptable case where is in `incremental` mode as well as `earliest`and `report duration` are present, why should we verify `lastestUpdatedAt` above?

Reply:
new `latestUpdatedAt` only be calculate if the task is incremental.